src/pcm.c

```diff
@@ -1641,6 +1641,17 @@ static int pcm_rw_transfer(struct pcm *pcm, void *data, unsigned int frames)
     transfer.frames = frames;
     transfer.result = 0;
 
+    /*
+     * If frames < start_threshold, wait indefinitely.
+     * Another thread may start capture
+     */
+    if (!is_playback && pcm_state(pcm) == PCM_STATE_PREPARED &&
+            frames >= pcm->config.start_threshold) {
+        if (pcm_start(pcm) < 0) {
```

Comment:
According to the ALSA kernel, it will start the stream. https://elixir.bootlin.com/linux/latest/source/sound/core/pcm_lib.c#L2294

Could you fill in more details?

Reply:
Hi David,
    
     That could be the case for kernel, but we are using pcm plugin instead so the stream needs to be started by tinyalsa.
     Based on our testing result, if not in MMAP mode, the stream will not be started

     Thanks,
     Huicheng

Reply:
Thanks for the background. In my opinion, the plugin should work like the ALSA kernel. Could you fix this on the plugin side?

Reply:
Hi David, 

Let me explain it more detail, this is the call flow with the current tinyalsa code:

playback:
**MMAP: open->prepare->start->write->stop->close**
**NON MMAP: open->prepare->write->close**

capture:
**MMAP: open->prepare->start->write->stop->close**
NON MMAP: open->prepare->write->close

With the above call flow, the three bold test case above is working good. You might be curious why playback NON MMAP mode doesn't give us any issue since it's also not calling pcm_start.
That is because in the function pcm_plug_writei_frames
https://github.com/tinyalsa/tinyalsa/blob/master/src/pcm_plugin.c#L549

static int **pcm_plug_writei_frames**(struct pcm_plug_data *plug_data,
                struct snd_xferi *x)
{
    struct pcm_plugin *plugin = plug_data->plugin;

    if (plugin->state != PCM_PLUG_STATE_PREPARED &&
        plugin->state != PCM_PLUG_STATE_RUNNING)
        return -EBADFD;

    return plug_data->ops->writei_frames(plugin, x);
}

at line 554 https://github.com/tinyalsa/tinyalsa/blob/master/src/pcm_plugin.c#L554, it checks if the plugin state is **PCM_PLUG_STATE_PREPARED or PCM_PLUG_STATE_RUNNING**. In this case it's ok to only call pcm_prepare but not calling pcm_start before pcm_write.

However the function pcm_plug_readi_frames https://github.com/tinyalsa/tinyalsa/blob/master/src/pcm_plugin.c#L561 is different.

static int pcm_plug_readi_frames(struct pcm_plug_data *plug_data,
                struct snd_xferi *x)
{
    struct pcm_plugin *plugin = plug_data->plugin;

    if (plugin->state != PCM_PLUG_STATE_RUNNING)
        return -EBADFD;

    return plug_data->ops->readi_frames(plugin, x);
}

at line 566 https://github.com/tinyalsa/tinyalsa/blob/master/src/pcm_plugin.c#L566, we have to make sure the plugin state is PCM_PLUG_STATE_RUNNING otherwise the function returns with failure. This means in capture case, we must call pcm_start after pcm_prepare, otherwise pcm_read will always fail.

We tried a workarond by changing the pcm state to PCM_PLUG_STATE_RUNNING in our plugin code while pcm_prepare is called but pcm_plug_prepare https://github.com/tinyalsa/tinyalsa/blob/master/src/pcm_plugin.c#L583 will change the state back to PCM_PLUG_STATE_PREPARED at line https://github.com/tinyalsa/tinyalsa/blob/master/src/pcm_plugin.c#L593.

static int pcm_plug_prepare(struct pcm_plug_data *plug_data)
{
    struct pcm_plugin *plugin = plug_data->plugin;
    int rc;

    if (plugin->state != PCM_PLUG_STATE_SETUP)
        return -EBADFD;

    rc = plug_data->ops->prepare(plugin);
    if (!rc)
        plugin->state = PCM_PLUG_STATE_PREPARED;

    return rc;
}

If this change is not accepted, then I'm thinking if we could make the pcm_plug_readi_frames function similar as pcm_plug_writei_frames by changing the if statement from https://github.com/tinyalsa/tinyalsa/blob/master/src/pcm_plugin.c#L566 to https://github.com/tinyalsa/tinyalsa/blob/master/src/pcm_plugin.c#L554.
Can you also let me know the reason why the if statement is different between these two functions? From pcm_plugin.c it looks like for capture NON MMAP case, pcm_start is expected to be called by tinyalsa.

We are ok if the pcm_start is not called in tinyalsa, we can change our plugin code. The real blocking issue is the plugin state here.

Reply:
I really appreciate your help to gather all the background of this issue. I think that is really helpful!

My only concern is if the ALAS kernel starts the capture internally, whether we can also do that in the pcm_plugin.c to align the behaviors between the ALSA kernel and the plugin? If we add a specific code for plugin cases, I think that might make others confused since most of the ALSA clients use the ALSA kernel as the underlying layer.

> If this change is not accepted, then I'm thinking if we could make the pcm_plug_readi_frames function similar as pcm_plug_writei_frames by changing the if statement from https://github.com/tinyalsa/tinyalsa/blob/master/src/pcm_plugin.c#L566 to https://github.com/tinyalsa/tinyalsa/blob/master/src/pcm_plugin.c#L554.
Can you also let me know the reason why the if statement is different between these two functions? From pcm_plugin.c it looks like for capture NON MMAP case, pcm_start is expected to be called by tinyalsa.

Actually I have no idea why the statement is different between these two functions. The plugin wrapper was also contributed by another authors. However, I think we should make the behavior more consistent between the ALSA kernel and the plugin wrapper. In my opinion, your idea to fix this issue on the plugin wrapper side is more reasonable. Could you please help to make a new change for this fix?

Reply:
Hi David, I have made the fix again. Unfortunately the commit history is messed up during force push of the github GUI. I created a new PR https://github.com/tinyalsa/tinyalsa/pull/261 and mentioned the current PR in the desciption.